clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -206,11 +209,21 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
+    }
+
+    public void closeWithTransactionState(long pid, short epoch, int baseSequence) {
+        recordsBuilder.closeWithTransactionState(pid, epoch, baseSequence);
+        isWritable = false;
```

Comment:
Just explaining my observations here: the only place we were not closing the builder while setting it to false at the same time is line 80 above. And if we want to reflect that a previous tryAppend has failed in `isWritable` we close just close the batch as we did before after `tryAppend` failed.